node/actors/network/src/http/mod.rs

```diff
@@ -0,0 +1,280 @@
+                                        tracing::error!("Error serving connection: {:?}", e);
+                                    }
+
+                                    Ok(())
+                                });
+                            },
+                            Err(err) => {
+                                if let Some(remote_addr) = err.peer_addr() {
+                                    tracing::error!("[client {remote_addr}] ");
+                                }
+
+                                tracing::error!("Error accepting connection: {}", err);
+                            }
+                        }
+                    },
+
+                    _ = &mut cancel_signal => {
+                        tracing::info!("canceled");
+
+                        // Now start the shutdown and wait for them to complete
+                        // Optional: start a timeout to limit how long to wait.
+                        graceful.shutdown().await;
+                        return Ok(())
+                    }
+                }
+            }
+        })
+        .await
+    }
+
+    async fn handle(
+        &self,
+        request: Request<hyper::body::Incoming>,
+    ) -> anyhow::Result<Response<Full<Bytes>>> {
+        let mut response = Response::new(Full::default());
+        match self.basic_authentication(request.headers()) {
+            Ok(_) => *response.body_mut() = self.serve(request),
+            Err(e) => {
+                *response.status_mut() = StatusCode::UNAUTHORIZED;
+                *response.body_mut() = Full::new(Bytes::from(format!("{}", e)));
+                let header_value = HeaderValue::from_str(r#"Basic realm="debug""#).unwrap();
+                response
+                    .headers_mut()
+                    .insert(header::WWW_AUTHENTICATE, header_value);
+            }
+        }
+        Ok(response)
+    }
+
+    fn basic_authentication(&self, headers: &HeaderMap) -> anyhow::Result<()> {
+        self.credentials.clone().map_or(Ok(()), |credentials| {
+            // The header value, if present, must be a valid UTF8 string
+            let header_value = headers
+                .get("Authorization")
+                .context("The 'Authorization' header was missing")?
+                .to_str()
+                .context("The 'Authorization' header was not a valid UTF8 string.")?;
+            let base64encoded_segment = header_value
+                .strip_prefix("Basic ")
+                .context("The authorization scheme was not 'Basic'.")?;
+            let decoded_bytes = base64::engine::general_purpose::STANDARD
+                .decode(base64encoded_segment)
+                .context("Failed to base64-decode 'Basic' credentials.")?;
+            let incoming_credentials = String::from_utf8(decoded_bytes)
+                .context("The decoded credential string is not valid UTF8.")?;
+            if Into::<String>::into(credentials) == incoming_credentials {
```

Comment:
`String::from`, but you should prefer comparing structured data instead (not strings).